src/Http/WebUtilities/src/MultipartPipeReaderStream.cs

```diff
@@ -0,0 +1,327 @@
+                        sequence = sequence.Slice(read);
+                        return (false, read);
+                    }
+                    if (body.Length > 0)
+                    {
+                        //copy all bytes until persumed boundary
+                        read += CopySequenceToBuffer(body, buffer, offset + read);
+                    }
+
+                    //there are more bytes than the received buffer, so return and slice bufferedData.
+                    if (read == buffer.Length - offset)
+                    {
+                        sequence = sequence.Slice(read);
+                        return (false, read);
+                    }
+
+                    //continue to check for boundaryMatch
+                    for (int i = _partialMatchIndex + 1; i < _boundary.BoundaryBytes.Length; i++)
+                    {
+                        if (sequenceReader.TryRead(out var value))
+                        {
+                            if (_boundary.BoundaryBytes[i] == value)
+                            {
+                                continue;
+                            }
+                            else
+                            {
+                                //no match - copy previous potential matches
+                                int j = 0;
+                                for (int bufferIndex = offset + read; bufferIndex < buffer.Length && j < i; bufferIndex++)
+                                {
+                                    buffer[bufferIndex] = _boundary.BoundaryBytes[j];
+                                    j++;
+                                }
+                                read += j;
+                                _partialMatchIndex = 0;
+                                sequence = sequence.Slice(sequenceReader.Consumed - 1);
+                                return (false, read);
+                            }
+                        }
+                        else
+                        {
+                            if (sequenceReader.End)
+                            {
+                                //might be a partial Match, need to read more to know for sure
+                                _partialMatchIndex = i;
+                                sequence = sequence.Slice(read);
+                                //slice partial read to insure _bufferedData is empty in the next check
+                                sequence = sequence.Slice(i);
+                                return (false, read);
+                            }
+                        }
+                    }
+                    sequence = sequence.Slice(_boundary.BoundaryBytes.Length - _partialMatchIndex);
+                    RawLength += _boundary.BoundaryBytes.Length;
+                    return (true, read);
+                }
+            }
+            return (false, read);
+        }
+
+        private bool TrySkipMetadata(ref ReadOnlySequence<byte> sequence, bool isFinalBlock)
```

Comment:
What is the metadata?

Reply:
The metadata is either the `--` or the new line after the boundary bytes. I didn't have a better name for it - it basically refers to this comment

```
// "The boundary may be followed by zero or more characters of
// linear whitespace. It is then terminated by either another CRLF"
// or -- for the final boundary.
```